### PR TITLE
Add category

### PR DESCRIPTION
**Pull request by teract10s**

### CODE REVIEW

File: src/main/java/com/example/bookstore/controller/CategoryController.java [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+@Tag(name = "Category management", description = "Endpoints for managing category")
+@RestController
+@RequiredArgsConstructor
+@RequestMapping("/api/categories")
+public class CategoryController {
+    private final CategoryService categoryService;
+    private final BookService bookService;
+
+    @GetMapping("")
```
Comment (NazarSmith228) [inline src/main/java/com/example/bookstore/controller/CategoryController.java]:
```suggestion
    @GetMapping
```

File: src/main/java/com/example/bookstore/controller/CategoryController.java [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+    @Operation(summary = "Get all categories")
+    @ResponseStatus(HttpStatus.OK)
+    public List<CategoryResponseDto> getAll(
+            @PageableDefault(page = 0, size = 10, sort = "name") Pageable pageable
+    ) {
```
Comment (NazarSmith228) [inline src/main/java/com/example/bookstore/controller/CategoryController.java]:
```suggestion
            @PageableDefault(page = 0, size = 10, sort = "name") Pageable pageable) {
```

File: src/main/java/com/example/bookstore/controller/CategoryController.java [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+        return categoryService.getAll(pageable);
+    }
+
+    @GetMapping("/{categoryId}/books")
+    @Operation(summary = "Get books by category",
+            description = "Getting books by receiving categoryId in path")
+    @ResponseStatus(HttpStatus.OK)
+    public List<BookResponseDto> getBooksByCategory(
+            @PathVariable Long categoryId
+    ) {
```
Comment (NazarSmith228) [inline src/main/java/com/example/bookstore/controller/CategoryController.java]:
```suggestion
    public List<BookResponseDto> getBooksByCategory(@PathVariable Long categoryId) {
```

File: src/main/java/com/example/bookstore/controller/CategoryController.java [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+        return bookService.getBookByCategoryId(categoryId);
+    }
+
+    @GetMapping("/{id}")
+    @Operation(summary = "Get category by id")
+    @ResponseStatus(HttpStatus.OK)
+    public CategoryResponseDto getCategory(@PathVariable Long id) {
```
Comment (NazarSmith228) [inline src/main/java/com/example/bookstore/controller/CategoryController.java:57]: will you validate `@PathVariable`?

if not - feel free to skip this comment

File: src/main/java/com/example/bookstore/controller/CategoryController.java [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+        return categoryService.getCategoryById(id);
+    }
+
+    @PostMapping("")
```
Comment (NazarSmith228) [inline src/main/java/com/example/bookstore/controller/CategoryController.java]:
```suggestion
    @PostMapping
```

File: src/main/java/com/example/bookstore/controller/CategoryController.java [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+    @Operation(summary = "Create new category")
+    @ResponseStatus(HttpStatus.CREATED)
+    @PreAuthorize("hasAuthority('ADMIN')")
+    public CategoryResponseDto createCategory(
+            @RequestBody @Valid CreateCategoryRequestDto categoryDto
+    ) {
```
Comment (NazarSmith228) [inline src/main/java/com/example/bookstore/controller/CategoryController.java]:
```suggestion
    public CategoryResponseDto createCategory(@RequestBody @Valid CreateCategoryRequestDto categoryDto) {
```

File: src/main/java/com/example/bookstore/controller/CategoryController.java [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+    }
+
+    @PutMapping("/{id}")
+    @Operation(summary = "Update category")
+    @ResponseStatus(HttpStatus.OK)
+    @PreAuthorize("hasAuthority('ADMIN')")
+    public CategoryResponseDto updateCategory(
+            @PathVariable Long id,
+            @RequestBody @Valid CreateCategoryRequestDto categoryDto
+    ) {
```
Comment (NazarSmith228) [inline src/main/java/com/example/bookstore/controller/CategoryController.java]:
```suggestion
    public CategoryResponseDto updateCategory(@PathVariable Long id,
            @RequestBody @Valid CreateCategoryRequestDto categoryDto) {
```

File: src/main/java/com/example/bookstore/dto/category/CategoryResponseDto.java [hunk 8]
```diff
@@ -0,0 +1,4 @@
+package com.example.bookstore.dto.category;
+
+public record CategoryResponseDto(Long id, String name, String description) {
```
Comment (NazarSmith228) [inline src/main/java/com/example/bookstore/dto/category/CategoryResponseDto.java]: please use the same formatting for all records

File: src/main/java/com/example/bookstore/mapper/BookMapper.java [hunk 9]
```diff
@@ -15,7 +15,7 @@
         nullValueCheckStrategy = NullValueCheckStrategy.ALWAYS,
         implementationPackage = "<PACKAGE_NAME>.impl")
 public interface BookMapper {
```
Comment (NazarSmith228) [inline src/main/java/com/example/bookstore/mapper/BookMapper.java:20]: shouldn't this mapper contain:
```
BookDtoWithoutCategoryIds toDtoWithoutCategories(Book book);
```
according to the task description

File: src/main/java/com/example/bookstore/controller/BookController.java [hunk 10]
```diff
@@ -33,35 +33,36 @@ public class BookController {
...
         return bookService.save(bookDto);
     }
 
     @PutMapping("/{id}")
     @Operation(summary = "Update book", description = "Update book by id")
     @ResponseStatus(HttpStatus.OK)
     @PreAuthorize("hasRole('ADMIN')")
-    public BookDto updateBook(Authentication authentication, @PathVariable Long id,
-                              @RequestBody CreateBookRequestDto createBookRequestDto) {
+    public BookResponseDto updateBook(Authentication authentication, @PathVariable Long id,
```
Comment (NazarSmith228) [inline src/main/java/com/example/bookstore/controller/BookController.java]: pls ensure you did everything according to requirements

should the book response contain `categoryIds`?

File: src/main/java/com/example/bookstore/model/Book.java [hunk 11]
```diff
@@ -40,4 +44,10 @@ public class Book {
 
     @Column(name = "is_deleted", nullable = false)
     private boolean isDeleted = false;
+
+    @ManyToMany()
```
Comment (NazarSmith228) [inline src/main/java/com/example/bookstore/model/Book.java]:
```suggestion
    @ManyToMany
```

File: src/main/java/com/example/bookstore/repository/CategoryRepository.java [hunk 12]
```diff
@@ -0,0 +1,7 @@
+package com.example.bookstore.repository;
+
+import com.example.bookstore.model.Category;
+import org.springframework.data.jpa.repository.JpaRepository;
+
+public interface CategoryRepository extends JpaRepository<Category, Long> {
```
Comment (NazarSmith228) [inline src/main/java/com/example/bookstore/repository/CategoryRepository.java:8]: `@Repository` missing

File: src/main/java/com/example/bookstore/service/impl/CategoryServiceImpl.java [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+        Category category = categoryRepository.findById(id)
+                .orElseThrow(() -> new EntityNotFoundException("Can't find category by id: " + id));
+        return categoryMapper.toDto(category);
+    }
+
+    @Override
+    public CategoryResponseDto save(CreateCategoryRequestDto categoryDto) {
+        Category category = categoryMapper.toCategory(categoryDto);
+        Category savedCategory = categoryRepository.save(category);
+        return categoryMapper.toDto(savedCategory);
```
Comment (NazarSmith228) [inline src/main/java/com/example/bookstore/service/impl/CategoryServiceImpl.java]: this can be inlined

**Review by NazarSmith228 — CHANGES_REQUESTED** [review]
Good job! Let's improve your solution

File: src/main/java/com/example/bookstore/controller/CategoryController.java (hunk 6 above)
Comment (teract10s) [reply src/main/java/com/example/bookstore/controller/CategoryController.java]: In this case maven build will fail. I have to have less than 100 chars in one line

File: src/main/java/com/example/bookstore/service/impl/BookServiceImpl.java [hunk 14]
```diff
@@ -41,7 +42,9 @@ public BookDto updateById(CreateBookRequestDto bookRequestDto, Long id) {
         Book bookFromDb = bookRepository.findById(id)
                 .orElseThrow(() -> new NoSuchElementException("Can't find book by id: " + id));
         bookMapper.updateBook(bookRequestDto, bookFromDb);
-        return bookMapper.toDto(bookRepository.save(bookFromDb));
+        BookDto dto = bookMapper.toDto(bookRepository.save(bookFromDb));
+        bookMapper.setCategoryIds(dto, bookFromDb);
```
Comment (NazarSmith228) [inline src/main/java/com/example/bookstore/service/impl/BookServiceImpl.java]: this method should be called automatically, since it is `@AfterMapping`

**Review by NazarSmith228 — APPROVED** [review]
LGTM, but please pay attention to my comment

**Pull request merged**